### PR TITLE
[SPARK-9788][MLlib] Fix LDA Binary Compatibility

### PR DESCRIPTION
**Pull request by feynmanliang**
1. Add “asymmetricDocConcentration” and revert docConcentration changes. If the (internal) doc concentration vector is a single value, “getDocConcentration" returns it. If it is a constant vector, getDocConcentration returns the first item, and fails otherwise.
2. Give `LDAModel.gammaShape` a default value in `LDAModel` concrete class constructors.

@jkbradley

### CODE REVIEW

**Comment by jkbradley** [comment]
reviewing now

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala [hunk 1]
```diff
@@ -79,7 +79,22 @@ class LDA private (
    *
    * This is the parameter to a Dirichlet distribution.
    */
-  def getDocConcentration: Vector = this.docConcentration
+  def getAsymmetricDocConcentration: Vector = this.docConcentration
+
+  /**
+   * Gets the concentration parameter, assuming the document-topic Dirichlet distribution is
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: Copy doc here:
> Concentration parameter (commonly named "alpha") for the prior placed on documents' distributions over topics ("theta").

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala [hunk 2]
```diff
@@ -79,7 +79,22 @@ class LDA private (
...
+   * symmetric. Included for backwards compatibility. This method should fail if
+   * [[docConcentration]] is asymmetric.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: "docConcentration" is linking to a private var.  Remove brackets.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala [hunk 3]
```diff
@@ -110,14 +125,17 @@ class LDA private (
     this
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: As long as you're editing here, can you add a check for docConcentration.length > 0?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala [hunk 4]
```diff
@@ -190,7 +189,8 @@ class LocalLDAModel private[clustering] (
     val topics: Matrix,
     override val docConcentration: Vector,
     override val topicConcentration: Double,
-    override protected[clustering] val gammaShape: Double) extends LDAModel with Serializable {
+    override protected[clustering] val gammaShape: Double = 100
+  ) extends LDAModel with Serializable {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: style: closing parenthesis should go on previous line

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala [hunk 5]
```diff
@@ -455,8 +455,9 @@ class DistributedLDAModel private[clustering] (
     val vocabSize: Int,
     override val docConcentration: Vector,
     override val topicConcentration: Double,
-    override protected[clustering] val gammaShape: Double,
-    private[spark] val iterationTimes: Array[Double]) extends LDAModel {
+    private[spark] val iterationTimes: Array[Double],
+    override protected[clustering] val gammaShape: Double = 100
+  ) extends LDAModel {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: style: closing parenthesis should go on previous line

**Comment by jkbradley** [comment]
Note (not requiring changes): I had meant for the default gammaShape to be defined in the abstract LDAModel, in order to improve (but not fix) binary incompatibility issues for anyone who extended LDAModel.  However, I just noticed that LDAModel has a private constructor, so we don't need to worry about binary incompatibility.  So feel free to arrange gammaShape however you like.  I think the current setup is fine.

**Comment by jkbradley** [comment]
This will need an update to PySpark as well.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala (hunk 1 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala (hunk 2 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala (hunk 3 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDA.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala (hunk 5 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: OK

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by jkbradley** [comment]
(Are you updating PySpark in a separate PR?)

**Comment by feynmanliang** [comment]
@jkbradley This PR doesn't need a PySpark update. Eventually, `PythonMLLibAPI.scala` will need to be updated to support asymmetric priors, hyperparam optimizations, and the predict methods. However, I think it might be better to hold off until the EM optimizer and Online optimizer have a consistent API.

**Comment by jkbradley** [comment]
True, it's outside this PR.  OK we'll see if tests will ever pass

**Comment by jkbradley** [comment]
Jenkins test this please

**Comment by feynmanliang** [comment]
Jenkins test this please

**Comment by jkbradley** [comment]
Merging with master and branch-1.5

**Pull request closed without merging**